src/tz_macos.rs

```diff
@@ -1,22 +1,61 @@
-use core_foundation_sys::base::{CFRelease, CFTypeRef};
-use core_foundation_sys::string::{kCFStringEncodingUTF8, CFStringGetCStringPtr};
+use core_foundation_sys::base::{Boolean, CFRange, CFRelease, CFTypeRef};
+use core_foundation_sys::string::{
+    kCFStringEncodingUTF8, CFStringGetBytes, CFStringGetCStringPtr, CFStringGetLength,
+};
 use core_foundation_sys::timezone::{CFTimeZoneCopySystem, CFTimeZoneGetName};
 
 pub(crate) fn get_timezone_inner() -> Result<String, crate::GetTimezoneError> {
-    unsafe {
-        Dropping::new(CFTimeZoneCopySystem())
-            .and_then(|tz| Dropping::new(CFTimeZoneGetName(tz.0)))
-            .and_then(|name| {
-                let name = CFStringGetCStringPtr(name.0, kCFStringEncodingUTF8);
-                if name.is_null() {
-                    None
-                } else {
-                    Some(name)
-                }
-            })
-            .and_then(|name| std::ffi::CStr::from_ptr(name).to_str().ok())
-            .map(|name| name.to_owned())
-            .ok_or(crate::GetTimezoneError::OsError)
+    unsafe { get_timezone().ok_or(crate::GetTimezoneError::OsError) }
+}
+
+#[inline]
+unsafe fn get_timezone() -> Option<String> {
+    // The longest name in the IANA time zone database is 25 ASCII characters long.
+    const MAX_LEN: usize = 32;
+
+    // Get system time zone, and borrow its name.
+    let tz = Dropping::new(CFTimeZoneCopySystem())?;
+    let name = CFTimeZoneGetName(tz.0);
+    if name.is_null() {
+        return None;
+    }
+
+    // If the name is encoded in UTF-8, copy it directly.
+    let cstr = CFStringGetCStringPtr(name, kCFStringEncodingUTF8);
+    if !cstr.is_null() {
+        let cstr = std::ffi::CStr::from_ptr(cstr);
+        if let Ok(name) = cstr.to_str() {
+            return Some(name.to_owned());
+        }
+    }
+
+    // Otherwise convert the name to UTF-8.
+    let mut buf = [0; MAX_LEN];
```

Comment:
I love that we're avoiding the heap allocation here but would like to make sure our assumptions are valid. What do you think about adding an assert that `MAX_LEN` is sufficient? If the assert triggers, we at least get a shot at getting a bug report.

`core-foundation` has a similar assert for the vector they allocate and abseil grabs the length to resize a vector:

https://github.com/servo/core-foundation-rs/blob/e64eab3ecad0071aff2f08aa4254f797a85475d1/core-foundation/src/string.rs#L137-L141

https://github.com/servo/core-foundation-rs/blob/e64eab3ecad0071aff2f08aa4254f797a85475d1/core-foundation/src/string.rs#L57-L68

I think the `core-foundation` crate is likely to be more correct given it doesn't make any assumptions about the `CFString`'s encoding.

https://github.com/abseil/abseil-cpp/blob/59cba2b50dd689010f338c63db95ca164f195798/absl/time/internal/cctz/src/time_zone_lookup.cc#L142-L143

Reply:
oh I see now, the check below for `buf_bytes` being in range checks for this. All good.